Address Gorry's review on -18 (#58)
core-wg/draft-dns-over-coap: draft-ietf-core-dns-over-coap.md

The term "constrained nodes" is used as defined in {{-constr-nodes}}.
{{-core-link-format}} describes that "Constrained RESTful Environments (CoRE)" realize the Representational State Transfer (REST) architecture {{REST}} in a suitable form for such constrained nodes.

> If the DoC server provides Observe capabilities, it is a "subject" as defined in {{-coap-observe, Section 1.2}}.
> This includes administering a "list of observers".

Reviewer: ```suggestion
A DoC server can provide Observe capabilities as defined in {{-coap-observe, Section 1.2}}.
As part of that, it administers a "list of observers".
```

The "subject" term is present in RFC7641, but I have never encountered it in any later literature, implementations, or spoken use.

DoC clients using these capabilities are "observers" as defined in {{-coap-observe, Section 1.2}} in that case.

> A "notification" is a CoAP response message with an Observer option, see {{-coap-observe, Section 4.2}}.

Reviewer 2: @miri64 I think I've spotted a small typo here:

```suggestion
A "notification" is a CoAP response message with an Observe option, see {{-coap-observe, Section 4.2}}.
```

However, it results in additional overhead, which conflicts with constrained resources.
This is the reason why it is RECOMMENDED to use CoAP Observe {{-coap-observe}} instead of DNS Push
in the DoC domain.
This is particularly useful if a short-lived record is requested frequently.
The DoC server SHOULD provide Observe capabilities on its DoC resource and do as follows.

> If a DoC clients wants to observe a resource record, a DoC server acts as subject and
> adds the client to its list of observers for that resource in accordance to {{-coap-observe}}.

Reviewer: ```suggestion
If a DoC clients wants to observe a resource record, a DoC server can respond with a notification
and add the client to its list of observers for that resource in accordance to {{-coap-observe}}.
```

See above on "subject"

The DoC server MAY subscribe to DNS push notifications for that record.
This involves sending a DNS Subscribe message (see ({{Section 6.2 of -dns-push}}),
instead of a classic DNS query to fetch the
information on behalf of the DoC client.

After the list of observers for a particular DNS query has become empty
(by explicit or implicit cancellation of the observation as per {{Section 3.6 of -coap-observe}}),
and no other reason to subscribe to that request is present,
the DoC server SHOULD cancel the corresponding subscription.
This can involve sending an DNS Unsubscribe message or closing the session (see {{Section 6.4 of -dns-push}}).
As there is no CoAP observer anymore from the perspective of the DoC server, a failure to do so cannot be communicated back to any DoC observer.
As such, error handling (if any) needs to be resolved between the DoC server and the upstream DNS infrastructure.

Whenever the DoC server receives a DNS Push message from the DNS
infrastructure for an observed resource record, the DoC server sends an appropriate Observe notification response
to the DoC client.

> A server that cannot proxy to DNS Push (or any other means of obtaining new records) SHOULD respond with a plain response message (i.e., without the Observe option) to a registration attempt.
> Observe notification for other purposes, however, are still possible.
> For instance, if the DoC server wants to notify about a change to a record it is the authoritative name server for, e.g., a new IP address for the name.

Reviewer: I think this'd cause more confusion than being helpful, esp. as to what the other purposes are -- and we're more in the factual requirements space here than in the normative SHOULD.

Maybe:

```suggestion
A server that responds with notifications (i.e., with the Observe option)
needs to have a means of obtaining current representations.
This may happend through DNS Push,
but also by polling
or implicitly (if the DoC is the authoritative name server for the request).
```

Author: This suggestion needs some fixes IMHO.

```suggestion
A server that responds with notifications (i.e., a response with the Observe option)
needs to have means of obtaining current resource records.
This may happen through DNS Push as described above,
but also by upstream polling
or through implicit circumstances (e.g., if the DoC is the authoritative name server for the record and wants to notify about changes).
```

Reviewer: Maybe a bit less verbose?

```suggestion
A server that responds with notifications (i.e., sends the Observe option)
needs to have means of obtaining current resource records.
This may happen through DNS Push,
but also by upstream polling
or through implicit circumstances (e.g., if the DoC is the authoritative name server for the request).
```

Author: Then let's do a bit less verbose, but also a bit more precise version ;-)

```suggestion
A server that responds with notifications (i.e., sends the Observe option) needs to have means of obtaining current resource records.
This may happen through DNS Push, but also by upstream polling or implicit circumstances (e.g., if the DoC server is the authoritative name server for the record and wants to notify about changes).
```
